python-scripts/eeprom-util.py

```diff
@@ -0,0 +1,198 @@
+# from the Eeprom method object (eeprom.hddKey + eeprom.gameReg). This is also why we
+# don't use any standard RC4 crypto libaries here.
+
+class EepKeyObj:
+  state = bytearray(range(256))
+  i = 0
+  j = 0
+
+def eepKeyPrepCrypt(eepKey, eepKeyObj):
+  j = 0
+
+  for i in range(256):
+    j = (j + eepKeyObj.state[i] + eepKey[i % len(eepKey)]) % 256
+    eepKeyObj.state[i], eepKeyObj.state[j] = eepKeyObj.state[j], eepKeyObj.state[i]
+
+def eepKeyCrypt(eepCryptData, eepKeyObj):
+  eepData = bytearray()
+  i = eepKeyObj.i
+  j = eepKeyObj.j
+  xorInd = 0
+
+  for byte in eepCryptData:
+    i = (i + 1) % 256
+    j = (j + eepKeyObj.state[i]) % 256
+    eepKeyObj.state[i], eepKeyObj.state[j] = eepKeyObj.state[j], eepKeyObj.state[i]
+    eepData.append(byte ^ eepKeyObj.state[(eepKeyObj.state[i] + eepKeyObj.state[j]) % 256])
+
+  eepKeyObj.i = i
+  eepKeyObj.j = j
+  return eepData
+
+parser = optparse.OptionParser()
+parser.add_option('-f', action='store', dest='eepFileLoc')
+parser.add_option('-k', action='store', dest='keyFileLoc')
+(options, args) = parser.parse_args()
+
+if options.eepFileLoc:
+  eeprom = Eeprom(options.eepFileLoc)
+  print ('\033[1mEncrypted EEPROM Data\033[0m')
+  print ('EEPROM Hash:\t\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.hash)))
+  print ('Encrypted Confounder:\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.confounder)))
+  print ('Encrypted HDD Key:\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.hddKey)))
+  print ('Encrypted Game Region Code:\t{}'.format(''.join(format(x, '02x') for x in eeprom.gameReg)))
+
+  print ('\n\033[1mIdentity EEPROM Data\033[0m')
+  print ('Identity Checksum:\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.identChk)))
+  print ('Serial Number:\t\t\t{}'.format(eeprom.serial.decode('ascii')))
+  print ('MAC Address:\t\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.mac)))
+  print ('Online Key:\t\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.onlineKey)))
+
+  # print NTSC / PAL instead of their respective codes
+  if ''.join(format(x, '02x') for x in eeprom.videoStd) == '00014000':
+    print ('Video Standard:\t\t\tNTSC')  
+  elif ''.join(format(x, '02x') for x in eeprom.videoStd) == '00038000':
+    print ('Video Standard:\t\t\tPAL')
+  else:
+    print ('Video Standard:\t\t\t{}'.format(''.join(format(x, '02x') for x in eeprom.videoStd)))
+
+if options.eepFileLoc and options.keyFileLoc:
+  # load eeprom key
+  eepKeyFile = open(options.keyFileLoc, 'rb')
+  eepKey = eepKeyFile.read()
+  
+  # generate hmac-sha1 hash of eeprom key
+  eepKeyHash = hmac.new(eepKey, eeprom.hash, sha1).digest()
+
+  # initialize EepKeyObj method object
+  eepKeyObj = EepKeyObj()
+  eepKeyPrepCrypt(eepKeyHash, eepKeyObj)
+  
+  # decrypt confounder 
+  decConf = eepKeyCrypt(eeprom.confounder, eepKeyObj)
+  
+  # assemble encrypted HDD key and game region code into a single bytearray
```

Comment:
yes, but isn't it easier to decrypt or hash these fields *before* unpacking them with read(4) etc.?